.github/workflows/tests.yml

```diff
@@ -0,0 +1,423 @@
+  # github repo settings.
+  TWINE_USERNAME: __token__
+
+  ###
+  # caching
+  # This is where we'd set up ccache, but this compiles so fast its not worth it.
+  ###
+
+
+
+jobs:
+  # Because sharing code/steps is so hard, and because it can be
+  # extremely valuable to be able to get binary wheels without
+  # uploading to PyPI and even if there is some failure, (e.g., for
+  # other people to test/debug), the strategy is to divide the process
+  # into several different jobs. The first builds and saves the binary
+  # wheels. It has dependent jobs that download and install the wheel
+  # to run tests, build docs, and perform linting. Building the
+  # manylinux wheels is an independent set of jobs.
+  #
+  # This divisin is time-saving for projects that take awhile to
+  # build, but somewhat less of a clear-cut win given how quick this
+  # is to compile (at least at this writing).
+  build-zope_interface:
+    # Sigh. Note that the matrix must be kept in sync
+    # with `test`, and `docs` must use a subset.
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        python-version: [2.7, pypy-2.7, pypy-3.6, 3.5, 3.6, 3.7, 3.8, 3.9]
+        os: [ubuntu-20.04, macos-latest]
+        exclude:
+          - os: macos-latest
+            python-version: pypy-2.7
+          - os: macos-latest
+            python-version: pypy-3.6
+          - os: macos-latest
+            python-version: 3.5
+
+    steps:
+      - name: checkout
+        uses: actions/checkout@v2
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+      ###
+      # Caching.
+      # This actually *restores* a cache and schedules a cleanup action
+      # to save the cache. So it must come before the thing we want to use
+      # the cache.
+      ###
+      - name: Get pip cache dir
+        id: pip-cache
+        run: |
+          echo "::set-output name=dir::$(pip cache dir)"
+
+      - name: pip cache
+        uses: actions/cache@v2
+        with:
+          path: ${{ steps.pip-cache.outputs.dir }}
+          key: ${{ runner.os }}-pip-${{ matrix.python-version }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
+
+      - name: Install Build Dependencies
+        run: |
+          pip install -U pip
+          pip install -U setuptools wheel twine
+          pip install -U coveralls coverage
+
+      - name: Build zope.interface
+        run: |
+          # Next, build the wheel *in place*. This helps ccache, and also lets us cache the configure
+          # output (pip install uses a random temporary directory, making this difficult).
+          python setup.py build_ext -i
+          python setup.py bdist_wheel
+          # Also install it, so that we get dependencies in the (pip) cache.
+          pip install -U coverage
+          pip install -U 'faulthandler; python_version == "2.7" and platform_python_implementation == "CPython"'
```

Comment:
wow, neat, first time I see environment markers used on the command line; I didn't know this was allowed!

Reply:
It's pretty cool, isn't it? The pip command line is just like `requirements.txt` or `install_requires=` in that it works in terms of abstract "requirements,"  which include all that stuff about environment markers.